File: experiments/grouped_phon/scripts/stats/extract_similarity.R
extract_similarity = function() {

    n_items = 6

    file_name = file.path("results", "long.csv")
    df = read.csv(file_name)

    models = c("slot_coding", "levenshtein", "syllable", "closed_biphone", paste0("open_biphone_", 1:5))

    all_values = data.frame()
    for (m in models) {

        print(m)

        temp = data.frame()

        all_sim = 0
        n_trials = length(levels(as.factor(df$trial)))
        for (i in 1:n_trials) {

            current_trial = df[df$trial == i,]
            current_trial = current_trial[1,]

            condition = current_trial$condition
            current = to_matrix(current_trial[[m]], n_items)
            
            if (condition == "grouped") {
                ind_1 = 1:3
                ind_2 = 4:6
            } else if (condition == "interleaved") {
                ind_1 = seq(1, n_items, 2)
                ind_2 = seq(2, n_items, 2)
            } else if (condition == "different") {
                ind = 1:n_items
            }

            if (condition == "different") {
                sim = get_sim(current, ind)
            } else {
                sim = mean(c(get_sim(current, ind_1), get_sim(current, ind_2)))
            }

            temp = rbind(temp, data.frame(metric = m, value = sim, condition = condition))

        }

        different = round(mean(temp$value[temp$condition == "different"]), digits = 3)
        grouped = round(mean(temp$value[temp$condition == "grouped"]), digits = 3)
        interleaved = round(mean(temp$value[temp$condition == "interleaved"]), digits = 3)

        all_values = rbind(all_values,
                           data.frame(Metric = m,
                                      Dissimilar = different,
                                      Grouped = grouped,
                                      Interleaved = interleaved))

    }

    file_name = file.path("results", "similarity_values.csv")
    write.csv(all_values, file_name, row.names = FALSE, quote = FALSE)

}

to_matrix = function(s, n) {
    matrix(as.numeric(strsplit(s, ',')[[1]]), n, n)
}

get_sim = function(m, ind) {
    sum = 0
    cnt = 0
    for (i in ind) {
        for (j in ind) {
            if (i != j) {
                sum = sum + m[i,j]
                cnt = cnt + 1
            }
        }
    }
    return(sum/cnt)
}

extract_similarity()

File: experiments/main_experiment/stimuli/list_generation/extract_similarity.jl
include("functions/to_include.jl")

function extract_similarity()

    all_metrics = ["slot_coding", "levenshtein", "syllable", "closed_biphone", "open_biphone_1"]

    file_name = "../../../../databases/results/lexique_reduced.csv"
    lexicon = CSV.read(file_name, DataFrame)

    all_matrices = Dict()
    for m in all_metrics
        file_name = "../../../../building/results/" * m * ".jld2"
        all_matrices[m] = load(file_name)["m"]
    end

    lists = CSV.read("results/lists.csv", DataFrame)
    for m in all_metrics

        lists[!,m] .= 0.0

        for i in 1:nrow(lists)

            current_list = string.(collect(lists[i,1:6]))
            condition = lists[i,"condition"]

            if condition == "similar"
                lists[i,m] = mean([
                    extract_values(current_list[1:3], lexicon, all_matrices[m]),
                    extract_values(current_list[4:6], lexicon, all_matrices[m])
                ])
            elseif condition == "dissimilar"
                lists[i,m] = extract_values(current_list, lexicon, all_matrices[m])
            else
                throw("Wrong condition definition")
            end

        end

    end

    df = DataFrame()
    for m in all_metrics

        temp = DataFrame()

        temp[!,"similarity_condition"] = ["similar", "dissimilar"]
        temp[!,"metric_condition"] .= m

        for m_sub in all_metrics

            ind_similar = lists.condition .== "similar" .&& lists.metric .== m
            ind_dissimilar = lists.condition .== "dissimilar" .&& lists.metric .== m

            temp[!,m_sub] .= [
                round(mean(lists[:,m_sub][ind_similar]), digits = 3);
                round(mean(lists[:,m_sub][ind_dissimilar]), digits = 3)]

        end

        append!(df, temp)

    end

    file_name = "results/similarity_values.csv"
    CSV.write(file_name, df)

    nothing

end

function extract_values(list, lexicon, matrix)

    n = length(list)

    list_phon = copy(list)
    for i in 1:n
        ind = findall(lexicon.ortho .== list[i])[1]
        list_phon[i] = lexicon.phon[ind]
    end

    sum = 0.0
    cnt = 0
    for i in 1:n
        for j in 1:n
            if i != j
                similarities[cnt]
                sum += matrix[String(list_phon[i]), String(list_phon[j])]
                cnt += 1
            end
        end
    end

    sum/cnt

end

extract_similarity()
